x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/saml/TransportSamlInvalidateSessionAction.java

```diff
@@ -102,44 +104,30 @@ private void findAndInvalidateTokens(SamlRealm realm, SamlLogoutRequestHandler.R
         final Map<String, Object> tokenMetadata = realm.createTokenMetadata(result.getNameId(), result.getSession());
         if (Strings.isNullOrEmpty((String) tokenMetadata.get(SamlRealm.TOKEN_METADATA_NAMEID_VALUE))) {
             // If we don't have a valid name-id to match against, don't do anything
-            logger.debug("Logout request [{}] has no NameID value, so cannot invalidate any sessions", result);
+            LOGGER.debug("Logout request [{}] has no NameID value, so cannot invalidate any sessions", result);
             listener.onResponse(0);
             return;
         }
 
         tokenService.findActiveTokensForRealm(realm.name(), containsMetadata(tokenMetadata), ActionListener.wrap(tokens -> {
-            logger.debug("Found [{}] token pairs to invalidate for SAML metadata [{}]", tokens.size(), tokenMetadata);
+            LOGGER.debug("Found [{}] token pairs to invalidate for SAML metadata [{}]", tokens.size(), tokenMetadata);
             if (tokens.isEmpty()) {
                 listener.onResponse(0);
             } else {
-                GroupedActionListener<TokensInvalidationResult> groupedListener = new GroupedActionListener<>(
-                    tokens.size(),
-                    ActionListener.wrap(collection -> listener.onResponse(collection.size()), listener::onFailure)
-                );
-                tokens.forEach(tuple -> invalidateTokenPair(tuple, groupedListener));
+                tokenService.invalidateAllTokens(tokens, ActionListener.wrap(tokensInvalidationResult -> {
+                    if (LOGGER.isInfoEnabled() && tokensInvalidationResult.getErrors().isEmpty() == false) {
+                        try (XContentBuilder builder = XContentBuilder.builder(XContentType.JSON.xContent())) {
+                            tokensInvalidationResult.toXContent(builder, ToXContent.EMPTY_PARAMS);
+                            LOGGER.info("Failed to invalidate some SAML access or refresh tokens {}", Strings.toString(builder));
+                        }
+                    }
+                    // return only the total of active tokens for users of the realm, i.e. not the number of actually invalidated tokens
+                    listener.onResponse(tokens.size());
+                }, listener::onFailure));
             }
         }, listener::onFailure));
     }
 
-    private void invalidateTokenPair(Tuple<UserToken, String> tokenPair, ActionListener<TokensInvalidationResult> listener) {
```

Comment:
Only an observation: before this change, we would invalidate tokens in pairs (refresh_0, access_0), ..., (refresh_n, access_n). Now we invalidate all refresh tokens first, then all access tokens. I can't think of why that'd be a problem and in fact it's probably desirable. Just jotting this down as a diff in behavior.

Reply:
Yes, I'm aware of this. I couldn't think of any reason why this change would matter in practice.